source/device_defender.c

```diff
@@ -3,30 +3,103 @@
  * SPDX-License-Identifier: Apache-2.0.
  */
 
+#include <aws/common/condition_variable.h>
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+#include <aws/common/zero.h>
 #include <aws/iotdevice/device_defender.h>
 #include <aws/iotdevice/external/cJSON.h>
+#include <aws/iotdevice/iotdevice.h>
 #include <aws/iotdevice/private/network.h>
 
 #include <aws/common/allocator.h>
+#include <aws/common/array_list.h>
+#include <aws/common/assert.h>
 #include <aws/common/atomics.h>
+#include <aws/common/byte_buf.h>
 #include <aws/common/clock.h>
+#include <aws/common/error.h>
 #include <aws/common/hash_table.h>
 #include <aws/common/logging.h>
+#include <aws/common/macros.h>
 #include <aws/common/string.h>
 #include <aws/common/task_scheduler.h>
+
 #include <aws/io/event_loop.h>
+
 #include <aws/mqtt/client.h>
+#include <aws/mqtt/mqtt.h>
+#include <string.h>
+
+/**
+ * Update s_copy_task_config() and aws_iotdevice_defender_config_destroy()
+ * when adding new members
+ */
+struct aws_iotdevice_defender_task_config {
+    struct aws_allocator *allocator;
+    struct aws_string *thing_name;
+    struct aws_array_list custom_metrics;
+    size_t custom_metrics_len;
+    enum aws_iotdevice_defender_report_format report_format;
+    uint64_t task_period_ns;
+    aws_iotdevice_defender_task_canceled_fn *task_canceled_fn;
+    aws_iotdevice_defender_task_failure_fn *task_failure_fn;
+    aws_iotdevice_defender_report_accepted_fn *accepted_report_fn;
+    aws_iotdevice_defender_report_rejected_fn *rejected_report_fn;
+    void *callback_userdata;
+};
+
+/**
+ * Instantiation of a custom metric that needs a value to be retrieved when
+ * it is time to produce a metric report.
+ */
+struct defender_custom_metric {
+    enum defender_custom_metric_type type;
+    struct aws_string *metric_name;
+    void *metric_cb_userdata;
+    union {
+        aws_iotdevice_defender_get_number_fn *get_number_fn;
+        aws_iotdevice_defender_get_number_list_fn *get_number_list_fn;
+        aws_iotdevice_defender_get_string_list_fn *get_string_list_fn;
+        aws_iotdevice_defender_get_ip_list_fn *get_ip_list_fn;
+    } supplier_fn;
+};
 
-struct aws_iotdevice_defender_v1_task {
+/**
+ * of a custom metric's data that needs to be populated into a report
```

Comment:
Did something get cut out here? Looks like some text is missing

Reply:
I only went so far back looking for the original text, so...just re-wrote the comment.